enhancements/hypershift/priority-class/priority-class.md

```diff
@@ -0,0 +1,266 @@
+- The CPO will reconcile the new override priority classes from the HCP spec, and propogate the them to the various components determined by the override criteria.
+  - The `DeploymentConfig` struct will be used as it is currently to propogate changes in priority classes.
+
+### maxSurge
+At the moment, we are mitigating a Kubernetes rollout bug that can take down the service by setting `maxSurge` to `0`. Due to the bug, we experience an outage if a pod is restarted for any reason during a rollout, and since overriding priority classes initiates a rollout, it could potentially lead to an outage. 
+
+As a temporary workaround, we've adjusted `maxSurge` values for all HA control plane deployments to `0`.
+PR: https://github.com/openshift/hypershift/issues/1568
+
+We'll need to re-evaluate if this needs to be reverted if/when the Kubernetes bug is resolved. 
+Kubernetes issue: https://github.com/openshift/hypershift/issues/1568 
+
+
+
+### Phase 2 - Dynamic Adjustment of Priority Classes (WIP)
+Phase 2 of this enhancement will focus on bolstering Hypershift's capability by introducing dynamic adjustment of priority classes based on woe workload's size. This will directly be tied to the HC's total node count.
+- Add new priority classes to the cmd/install assets for HyperShift Operator, and enable the utilization of these custom priority classes using a CLI flag. 
+  - Priority classes will be defined by a unique `Name` and `Value`. These will need to be read by external microservices, so the naming convention is important here. The `Name` must follow these general guidelines:
+    - contain no more than 253 characters
+    - contain only lowercase alphanumeric characters, '-' or '.'
+    - start with an alphanumeric character
+    - end with an alphanumeric character
+    - cannot be prefixed with `system-`
+  - The integer value of the priority class can be any 32-bit integer value smaller than or equal to 1 billion. Larger numbers are reserved for critical system Pods that should not normally be preempted or evicted. 
+    - Values are relatively assigned to priority classes based on management cluster, but it’s up to service provider to determine values. These values are not universally set and can differ between different customer enviroments.
+
+- Add logic for HO to calculate total node count for a specific HC:
+  - Within the Nodepool controller's reconcile functions, list all the nodepools. 
+  - Filter through nodepools to gather those associated with the hosted cluster in question. 
+  - Calculate the total node count for the cluster and override the priority class based on it will be added to `hypershift-operator/controllers/util`. 
+    - Assign a priority class by comparing the node count against the predetermined upper bounds of priority classes. 
+    - Override the priority class specification of the HC.
```

Comment:
The API should allow expressing whether a priority class setting comes from the external user of the API or the operator.

Reply:
For now we've implemented your suggestion of using a HC annotation instead, but will leave this marked unresolved in case we need to switch things up.

Reply:
>The API should allow expressing whether a priority class setting comes from the external user of the API or the operator.

Why would that semantic distinction be needed?